lib/rules/no-leaky-event-listeners.js

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+'use strict';
+
+const { docUrl } = require('../util/doc-url');
+
+module.exports = {
+    meta: {
+        docs: {
+            description: 'prevent event listeners from leaking memory',
+            category: 'LWC',
+            recommended: true,
+            url: docUrl('no-leaky-event-listeners'),
+        },
+
+        schema: [],
+    },
+
+    create(context) {
+        /**
+         * Returns true if the passed Identifier node is referencing a property on the global
+         * object.
+         */
+        function isIdentifierGlobal(identifier) {
+            let scope = context.getScope(identifier);
+
+            while (scope.upper) {
+                scope = scope.upper;
+            }
+
+            return scope.through.find(variable => variable.identifier === identifier);
+        }
+
+        /**
+         * Returns true of the passed string is a valid event listener method.
+         */
+        function isValidEventListenerMethodName(name) {
+            return name === 'addEventListener' || name === 'removeEventListener';
+        }
+
+        /**
+         * Emits if a given call expression dealing with an event listener that would leak.
+         */
+        function reportListenerLeak(callExpression, { target, method }) {
+            // Early exit if the considered EventTarget is not the document or the window. Event
+            // listeners added programmatically on a component element will automatically be
+            // garbage collected.
+            if (target !== 'window' && target !== 'document') {
```

Comment:
I would add `document.body` and `document.head` here. Although it's also possible to leak on e.g. an omnipresent header/footer/modal/etc element, i.e. anything that isn't `template`. So maybe just check if `template` is not in there?

Reply:
You can not only add event listeners on the shadow root using `this.template.addEventListener` but you can also add event listeners on the host element using `this.addEventListener`. Checking if the `template` property is not present might generate too many false positives

Instead of checking if the `template` is present or not in the member lookup expression, I would prefer checking if the event target is accessed via the `document` or the `window` object. Using this approach we should be able to catch the following patterns:

```js
document.addEventListener('test', handleTest.bind(this));
document.body.addEventListener('test', handleTest.bind(this));
document.querySelector('#my-modal').addEventListener('test', handleTest.bind(this));
```

But would miss the following patterns:

```js
const myModal = document.querySelector('#my-modal')
myModal.addEventListener('test', handleTest.bind(this));
```

Reply:
I updated the linting rule to support those case: https://github.com/salesforce/eslint-plugin-lwc/pull/34/files#diff-1dcb28173ad87790ba05ca49467def70R17-R24